server/asset/dcr/utxo.go

```diff
@@ -0,0 +1,121 @@
+		txOut, verboseTx, _, err := dcr.getTxOutInfo(&utxo.txHash, utxo.vout)
+		if err != nil {
+			return -1, err
+		}
+		// More than zero confirmations would indicate that the transaction has
+		// been mined. Collect the block info and update the utxo fields.
+		if txOut.Confirmations > 0 {
+			blk, err := dcr.getBlockInfo(verboseTx.BlockHash)
+			if err != nil {
+				return -1, err
+			}
+			utxo.height = uint32(blk.height)
+			utxo.blockHash = blk.hash
+		}
+	} else {
+		// The UTXO was included in a block, but make sure that the utxo's block has
+		// not been orphaned or voted as invalid.
+		mainchainBlock, found := dcr.blockCache.atHeight(utxo.height)
+		if !found {
+			return -1, fmt.Errorf("no mainchain block for tx %s at height %d", utxo.txHash.String(), utxo.height)
+		}
+		// If the UTXO's block has been orphaned, check for a new containing block.
+		if mainchainBlock.hash != utxo.blockHash {
+			// See if we can find the utxo in another block.
+			newUtxo, err := dcr.utxo(&utxo.txHash, utxo.vout)
+			if err != nil {
+				return -1, fmt.Errorf("utxo block is not mainchain")
+			}
+			*utxo = *newUtxo
+			if utxo.height == 0 {
+				mainchainBlock = nil
+			} else {
+				mainchainBlock, found = dcr.blockCache.atHeight(utxo.height)
+				if !found {
+					return -1, fmt.Errorf("new block not found for utxo moved from orphaned block")
+				}
+			}
+		}
+		// If the block is set, check for stakeholder invalidation.
+		if mainchainBlock != nil && !mainchainBlock.txInStakeTree(&utxo.txHash) {
+			nextBlock, err := dcr.getMainchainDcrBlock(utxo.height + 1)
+			if err != nil {
+				return -1, fmt.Errorf("error retreiving approving block for utxo %s:%d: %v", utxo.txHash, utxo.vout, err)
+			}
+			if nextBlock != nil && !nextBlock.vote {
+				return -1, fmt.Errorf("utxo's block has been voted as invalid")
+			}
+		}
+	}
+	// If the height is still 0, this is a mempool transaction.
+	if utxo.height == 0 {
+		return 0, nil
+	}
+	// Otherwise just check that there hasn't been a reorg which would render the
+	// output immature. This would be exceedingly rare (impossible?).
+	confs := int32(dcr.blockCache.tipHeight()) - int32(utxo.height) + 1
+	if confs < utxo.maturity {
+		return -1, fmt.Errorf("transaction %s became immature", utxo.txHash)
+	}
+	return int64(confs), nil
+}
+
+// PaysToPubkey verifies that the utxo pays to the supplied public key. This is
+// an asset.DEXAsset method. The second argument is for additional data as would
+// be needed if P2SH support was enabled.
+func (utxo *UTXO) PaysToPubkey(pubkey, _ []byte) bool {
+	pkHash := dcrutil.Hash160(pubkey)
+	extracted := extractPubKeyHash(utxo.pkScript)
+	return extracted != nil && bytes.Equal(extracted, pkHash)
+}
+
+// ScriptSize returns the maximum spend script size of the UTXO, in bytes.
+func (utxo *UTXO) ScriptSize() uint32 {
+	if isPubKeyHashScript(utxo.pkScript) {
+		return P2PKHSigScriptSize
+	}
+	// This condition should be impossible. If the UTXO was created, it has a
```

Comment:
Please note here that you mean if the UTXO was created by **`(*dcrBackend).utxo`**.  And reaching this condition indicates programmer error, I'm pondering if a panic might actually be warranted here.